### ISSUE TITLE
Lint warnings for web content developers/authors

### ISSUE THREAD

**Issue by matatk** [matatk/landmarks#182]
As suggested in 6ac112db998aaae568aaf0b47de2c6f7f9512b92, would it be helpful to emit warnings when invalid things such as `aria-labelledby=""` are encountered?

* Should it just do it, or should this be a separate mode?
* If the error has already been reported, should it be reported again (e.g. during a scan after a mutation)? (Probably not—could flag the first scan? But then that might miss the error, as it could be added by a mutation, so should a record of the errors be kept?)

**Update (2019/08/11):** I'm leaning towards it always linting, and exposing some icons in the pop-up at least (maybe even a flag on the icon?) In order to stop repeated results, it could take note of the selector computed for an element, as this should remain constant across certain page changes.

**Comment by matatk** [comment]
Some more lint suggestions (on top of the above):

* More than one "main" region.
* More than one region of *any* type that are lacking unique labels.

**Comment by carmacleod** [comment]
Note that the current [html spec for main element](https://html.spec.whatwg.org/multipage/grouping-content.html#the-main-element) is a bit different from the (old) W3C spec.

You can still only have one visible main element:
> A document must not have more than one main element that does not have the hidden attribute specified.

But the old spec said "main must not be a descendant of article, aside, footer, header or nav", which is similar to but not quite the same as:
> A hierarchically correct main element is one whose ancestor elements are limited to html, body, div, form without an accessible name, and autonomous custom elements.

**Comment by carmacleod** [comment]
@matatk Issue https://github.com/matatk/landmarks/issues/370 is a great example of where linting would be useful/helpful. :)

I like your updated idea of exposing some icons in the pop-up. Maybe (for issue 370 for example) that could look something like this? Add a tooltip "There should only be one visible main" for details?

![image](https://user-images.githubusercontent.com/3331913/83421895-c32cd900-a3f6-11ea-9f23-46ebf73cd4dd.png)

**Comment by matatk** [comment]
Thanks @carmacleod, for both the note on `<main>` and the UI suggestions; that looks quite like how it did when I imagined it :-). I'll need to think about whether these errors/warnings should be presented to the end-user (given there's nothing they can do about them) or just the developer via the DevTools panel (and maybe we might treat errors and warnings differently), but this is quite motivating—I want to see it in action now :-).

I have been doing quite a bit in preparation for this: I've been improving the profiling tools to make it easier to determine how adding the lining might affect performance, as well as researching the checks to perform.

This will still take some time, so I don't think it will be in the next release, as there are already quite a few fixes I'd like to get out, but I am working on it.

**Comment by carmacleod** [comment]
Sounds great, @matatk!

Re:
> I'll need to think about whether these errors/warnings should be presented to the end-user (given there's nothing they can do about them) 

Oh, right. My developer-centric brain was thinking that the developer was the end user, but of course, there are _real_ end users. :)  Sorry about that!

**Comment by matatk** [comment]
Finally started to address this, with just one simple check, in #410 :-).

**Comment by carmacleod** [comment]
> Lack of a main region?

Absolutely, flag lack of a main region, mostly because main is the most important landmark, but also if there's, say, a header and a footer and no main, then there must be a big chunk of:
> Content that is outside of a landmark region.

One exception that I noticed was google, which has only a single "search" landmark, and no main... which kind of makes sense...
(nothing else on the page is in a landmark, but it's a very sparse page)
